### DISCUSSION TITLE
sqlalchemy as sql builder feeding pysqlite

### DISCUSSION THREAD

**Discussion by loynoir** [sqlalchemy/sqlalchemy#7646] [Usage Questions] [answered]
### Describe the bug

named style went away when `compile` on `sqlite` dialect

### To Reproduce

```python
import sqlalchemy as sa
import sqlalchemy.dialects.sqlite as dialects_sqlite
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()

class Message(Base):
    __tablename__ = "message"
    id = sa.Column(sa.Integer, primary_key=True, autoincrement=True)
    message = sa.Column(sa.String)
    foo = sa.Column(sa.String, server_default="bar")


print(
    sa.insert(Message.__table__)
    .compile(dialect=dialects_sqlite.dialect())
)
```

### Error

```
INSERT INTO message (id, message, foo) VALUES (?, ?, ?)
``` 

### Expected
```
INSERT INTO message (id, message, foo) VALUES (:id, :message, :foo)
```

### Use case
Pre-compile sql to string, and pass to native python sqlite3 module, with named style params

### Workaround
```python
print(sa.insert(Message.__table__))
# INSERT INTO message (id, message, foo) VALUES (:id, :message, :foo)
```

### Versions

- OS:
- Python:
- SQLAlchemy: 1.4.31
- Database:
- DBAPI (eg: psycopg, cx_oracle, mysqlclient):

### Additional context

- `sqlite3` module support named style

https://docs.python.org/3/library/sqlite3.html

```python
# And this is the named style:
cur.execute("select * from lang where first_appeared=:year", {"year": 1972})
print(cur.fetchall())
```

**Comment by CaselIT** [comment, marked as answer]
Hi,

the sqlite dialect uses positional parameters so this is the expected behaviour.
You may be able to change it by replacing `paramstyle` of the dialect. Something like this
```py
sqlite_d = dialects_sqlite.dialect()
sqlite_d.paramstyle = "named"
```

Note that how you are compiling will not generate valid sql in all cases. You need to pass compiler args. See https://docs.sqlalchemy.org/en/14/faq/sqlexpressions.html?highlight=compiler_kwargs#how-do-i-render-sql-expressions-as-strings-possibly-with-bound-parameters-inlined

**Reply by loynoir** [reply]
Thanks! @CaselIT 

Now I can get generated sql like
```sql
INSERT INTO message (id, message, foo) VALUES (:id, :message, :foo)
```

How can I exclude columns with default values?

Given column `foo` has server default `bar`, expected below
```sql
-- INSERT INTO message (id, message, foo) VALUES (:id, :message, :foo)
INSERT INTO message (id, message) VALUES (:id, :message)
```

**Reply by loynoir** [reply]
<details>
  <summary>else on docs</summary>
  <p>
 
BTW, what keyword do you use on `https://docs.sqlalchemy.org` to get that link?

Even I know it now, I try with keyword `named faq`, results seems not helpful 😭
  </p>
</details>

**Reply by CaselIT** [reply]
I used `compile_kwargs`, but I agree that sphinx search is not always that good.

> How can I exclude columns with default values?

I don't think you can automatically. You would need to create an sqlalchemy insert statement with only the columns you need.
`sa.insert(Message.message, Message.foo)`

**Reply by loynoir** [reply]
@CaselIT 


ArgumentError: subject table for an INSERT, UPDATE or DELETE expected, got Column('message', String(), table=<message>) object resolved from <sqlalchemy.orm.attributes.InstrumentedAttribute object> object

**Reply by CaselIT** [reply]
I was going from memory.
It seems there may be an option to do what you want https://docs.sqlalchemy.org/en/14/core/dml.html#sqlalchemy.sql.expression.Insert.from_select.params.include_defaults

If that does not work I'm not sure how you can specify the columns without also passing values to the insert

**Reply by loynoir** [reply]
@CaselIT 
Kind of closer, but seems not same thing

### Expected
```sql
-- INSERT INTO message (id, message, foo) VALUES (:id, :message, :foo)
INSERT INTO message (id, message) VALUES (:id, :message)
```

### Current
```python
print(Message.__table__.insert().from_select(["message"], sa.select(Message.__table__)))
```
```sql
INSERT INTO message (message) SELECT message.id, message.message, message.foo 
FROM message
```

**Reply by loynoir** [reply]
Ha! Got it! Kind of!
```python
print(insert(Message.__table__).values(message=sa.bindparam('xxxxx')))
```

```sql
INSERT INTO message (message) VALUES (:xxxxx)
```

**Comment by loynoir** [comment]
Only insert neccessay column
```python
def insert_filter(cls, filter_func):
    d = {}
    for col in cls.__table__.columns:
        if filter_func(col):
            d[col.name] = sa.bindparam(col.name)
        else:
            pass

    return sa.insert(cls.__table__).values(**d)


def colHasDefault(col):
    return (
        getattr(col, "primary_key", False) and col.type.__class__.__name__ == "Integer"
    ) or getattr(col, "server_default", False)


def insert_bone(cls):
    return insert_filter(cls, filter_func=lambda col: not colHasDefault(col))


print(insert_bone(Message).compile(dialect=dialects_sqlite.dialect(paramstyle="named")))
# INSERT INTO message (message) VALUES (:message)
```

**Reply by CaselIT** [reply]
not the easiest code, but it seems to work fine.
@zzzeek do you have a better solution? The question is this one  https://github.com/sqlalchemy/sqlalchemy/discussions/7646#discussioncomment-2078265

>
>>
> >   How can I exclude columns with default values?
>>
>I don't think you can automatically. You would need to create an sqlalchemy insert statement with only the columns you need.

**Reply by zzzeek** [reply]
uh yeah, you can pass "paramstyle" to dialect and "column_keys" to compiler for insert

```py
stmt = Message.__table__.insert()

from sqlalchemy.dialects import sqlite

compiled = stmt.compile(
    dialect=sqlite.dialect(paramstyle="named"),
    column_keys=[
        c.key
        for c in Message.__table__.c
        if not c.server_default and not c.primary_key
    ],
)

print(compiled)
```

hopefully that's easy enough

**Reply by zzzeek** [reply]
you can generalize the autoincrement part like this:

```py
compiled = stmt.compile(
    dialect=sqlite.dialect(paramstyle="named"),
    column_keys=[
        c.key
        for c in Message.__table__.c
        if not c.server_default
        and c is not Message.__table__._autoincrement_column
    ],
)

```

**Reply by CaselIT** [reply]
Did not know about coluy_keys. Is it documented somewhere?

**Reply by zzzeek** [reply]
it hardly would be an interesting parameter if you could just find it in the **docs**.   

it's likely buried in a docstring for SQLCompiler, here it is: https://docs.sqlalchemy.org/en/14/core/internals.html?highlight=sqlcompiler#sqlalchemy.sql.compiler.SQLCompiler.params.column_keys

the use case here is an odd one, I dont think I've ever pointed anyone to this particular parameter :)
